runtime/cudaq/utils/tensor.h

```diff
@@ -0,0 +1,146 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2024 NVIDIA Corporation & Affiliates.                         *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+#pragma once
+
+#include "details/tensor_impl.h"
+#include "extension_point.h"
+#include "type_traits.h"
+
+namespace cudaqx {
+
+/// @brief A tensor class implementing the PIMPL idiom.
+template <typename Scalar = std::complex<double>>
+class tensor {
+private:
+  std::shared_ptr<details::tensor_impl<Scalar>> pimpl;
+
+public:
+  /// @brief Type alias for the scalar type used in the tensor
+  using scalar_type = details::tensor_impl<Scalar>::scalar_type;
```

Comment:
I think this might be needed for C++17 (or, more specifically, if we include this file in an nvcc compilation.
```suggestion
  using scalar_type = typename details::tensor_impl<Scalar>::scalar_type;
```